fix(ivy): throw meaningful error for uninitialized output

crisbeto opened the pull request:
Throws a similar error to ViewEngine when encountering an `@Output` that hasn't been initialized to an `Observable`.

These changes resolve FW-680.

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -885,10 +886,18 @@ export function listener(
+          lCleanup.push(listenerFn, subscription);
+          tCleanup && tCleanup.push(eventName, tNode.index, idx, -(idx + 1));
+        } else {
+          throw new Error(
```
Open question: could / should we do in the dev mode only in `ngIvy`?

crisbeto replied:
That was something I was considering as well, but I decided to keep it consistent with ViewEngine. I was also thinking about rewording the error a little bit, because currently it says that the value hasn't been initializied, whereas the check is actually that the value is an observable.

pkozlowski-opensource replied:
> That was something I was considering as well, but I decided to keep it consistent with ViewEngine

I don't know if there is much value in keeping it consistent here...

crisbeto replied:
Good point, although it could break somebody's tests in theory, if they were asserting the same error. That being said, it's a bit of a contrived example.

kara replied:
+1 on dev mode only

mary-poppins commented:
You can preview 47d3691 at https://pr28085-47d3691.ngbuilds.io/.

mary-poppins commented:
You can preview ffd9e20 at https://pr28085-ffd9e20.ngbuilds.io/.

crisbeto commented:
@pkozlowski-opensource @kara I've switched it to only be in dev mode, PTAL.

mary-poppins commented:
You can preview 44aa731 at https://pr28085-44aa731.ngbuilds.io/.

kara commented on packages/core/test/bundling/todo/bundle.golden_symbols.json:
```diff
@@ -920,6 +920,9 @@
     "name": "isNodeMatchingSelectorList"
   },
+  {
+    "name": "isObservable"
```
Seems like this should be tree-shaken away? Maybe re-run this?

crisbeto replied:
Yep looks like it did. I only ran the script once the first time I submitted the PR.

kara reviewed (approved):
LGTM

mary-poppins commented:
You can preview 96a16c3 at https://pr28085-96a16c3.ngbuilds.io/.

kara commented:
[presubmit](http://test/OCL:228918361:BASE:228919042:1547236100006:766914d3)

crisbeto committed:
fix(ivy): throw meaningful error for uninitialized output

Throws a similar error to ViewEngine when encountering an `@Output` that hasn't been initialized to an `Observable`.

These changes resolve FW-680.

mary-poppins commented:
You can preview 503b152 at https://pr28085-503b152.ngbuilds.io/.

mary-poppins commented:
You can preview 47d3691 at https://pr28085-47d3691.ngbuilds.io/.
You can preview ffd9e20 at https://pr28085-ffd9e20.ngbuilds.io/.
You can preview 44aa731 at https://pr28085-44aa731.ngbuilds.io/.
You can preview 96a16c3 at https://pr28085-96a16c3.ngbuilds.io/.
You can preview 503b152 at https://pr28085-503b152.ngbuilds.io/.

AndrewKushnir commented:
@crisbeto thanks for submitting this PR. After merging other changes to master branch there is now a conflict, so this PR can't be merged. Could you please rebase this PR from the latest master and put the "merge" label back (I've removed it for now)? Thank you.

crisbeto commented:
Rebased.

mary-poppins commented:
You can preview fd0dd80 at https://pr28085-fd0dd80.ngbuilds.io/.

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:229288515:BASE:229288565:1547516021267:39e0ade2)

mary-poppins commented:
You can preview fb5862c at https://pr28085-fb5862c.ngbuilds.io/.

AndrewKushnir commented:
@crisbeto I've rebased this PR on top of the latest master and run g3 presubmit again to verify everything looks good. Could you please have a quick look at the changes to make sure there are no unexpected changes after rebase and return "merge" label back if everything is ok? Thank you.
